### ISSUE TITLE
Legacy support: MBR with Grub

### ISSUE THREAD

**Issue by Torxed** [archlinux/archinstall#42]
Currently, one of the bigger limitations are that the installer/guided templates only support UEFI with [Systemd-boot](https://wiki.archlinux.org/index.php/Systemd-boot) as the default bootloader.

We'll need to add `archinstall.MBR` as an alternative to:

```py
with archinstall.Filesystem(harddrive, archinstall.GPT) as fs:
    ...
```

And `installation.add_bootloader()` should probably take a parameter for which bootloader to install. Currently it doesn't take any arguments, so adding one won't brake anything. Suggestion: `installation.add_bootloader("grub")`

**Comment by klausenbusk** [comment]
> We'll need to add `archinstall.MBR` as an alternative to:
> 
> ```python
> with archinstall.Filesystem(harddrive, archinstall.GPT) as fs:
> ```

MBR partitioning isn't required for "legacy boot", most BIOS works just fine with a GPT partitioned disk. A `BIOS boot` partition is required though, more info here: https://wiki.archlinux.org/index.php/GRUB#GUID_Partition_Table_(GPT)_specific_instructions

**Comment by grazzolini** [comment]
I don't recommend creating a frankenstein BIOS/GPT installation. Also, some firmware might not be able to boot GPT (although they are rare). So, go for UEFI/GPT or BIOS/MBR.

Advanced users might want to do the partitioning themselves.

**Comment by Torxed** [comment]
I've been trying to investigate a clean way of getting GPT with "lgacy boot", and I haven't come across any really good writeups about it. Most seam to either stick to GPT and use systemd-bootctl or grub *(even grub ones seam overly complicated in comparison to systemd-bootctl)*.

So I'm with you on this one @grazzolini, perhaps advanced users will settle for manual work, and perhaps the installer could be run after the fact and simply select a disk - detect there's a partition scheme going on - and install on a existing partition platform *(or simply, advanced users skip archinstall all together)*.

I won't exclude this feature request tho, if the need for this becomes bigger I'll bring focus to it again, but for now it'll remain dormant until it's a priority if that sounds good :)

**Comment by grazzolini** [comment]
Yes, this is something I've been thinking about proposing. The partitioning and filesystem can be done prior to running archinstall and in that case it would take whatever the user created and go with it. That way things are even more flexible and users have more choice.

**Comment by Torxed** [comment]
> Yes, this is something I've been thinking about proposing. The partitioning and filesystem can be done prior to running archinstall and in that case it would take whatever the user created and go with it. That way things are even more flexible and users have more choice.

I think that settles it, this is the way forward for this feature request.
It'll be prioritized after #56 is in place and some other features. But this part should be very easy to implement.

**Comment by grazzolini** [comment]
Thanks. Having a way to run archinstall from any given point is a killer feature, because one could mix the current way to install arch with archinstall and vice versa.

**Comment by advaithm** [comment]
I've started to work on this, its still pretty early but I think I might have a working implementation by the end of this week

**Comment by omeringen** [comment]
Sorry if this is a bit off topic but i am using arch on my USB stick. Since this setup works on multiple computers, i both have a 10MB partition for GRUB/BIOS (i used --removable option for grub-install) and 500MB partition for systemd-boot/UEFI. Rest is for root. Works fine.
It would be nice to have this kind of option on installer.

**Comment by Torxed** [comment]
> Sorry if this is a bit off topic but i am using arch on my USB stick. Since this setup works on multiple computers, i both have a 10MB partition for GRUB/BIOS (i used --removable option for grub-install) and 500MB partition for systemd-boot/UEFI. Rest is for root. Works fine.
> It would be nice to have this kind of option on installer.

I don't see why this couldn't be an option.
Essentially it's a portable boot-loader that you carry around by the sound of it.

**Comment by grazzolini** [comment]
I think this will add unneeded complexity to cover a very special and corner case. But hey, if it can be done without confusing users, by all means.

**Comment by advaithm** [comment]
I have to agree with @grazzolini I don't see a easy way to add this with out adding a lot of complexity for everyone.

**Comment by grazzolini** [comment]
@Torxed, MBR support, and other boot loaders support, on the other hand, is nice to have, for sure.

**Comment by thetayloredman** [comment]
> I've been trying to investigate a clean way of getting GPT with "lgacy boot", and I haven't come across any really good writeups about it. Most seam to either stick to GPT and use systemd-bootctl or grub _(even grub ones seam overly complicated in comparison to systemd-bootctl)_.
> 
> So I'm with you on this one @grazzolini, perhaps advanced users will settle for manual work, and perhaps the installer could be run after the fact and simply select a disk - detect there's a partition scheme going on - and install on a existing partition platform _(or simply, advanced users skip archinstall all together)_.
> 
> I won't exclude this feature request tho, if the need for this becomes bigger I'll bring focus to it again, but for now it'll remain dormant until it's a priority if that sounds good :)

I'd always create an 8MB grub_bios partition before the EFI-SP and use GRUB.

**Comment by grazzolini** [comment]
@thetayloredman you're wasting 8MB then, because grub stage 2 is hardcoded to be **at maximum** 31Kb. Usually the recommendation to create 1MB bios boot partition is not because grub needs that whole 1MB, it's a small partition that aligns with the physical disk sectors in most cases. However, grub will **never** use that much.

**Comment by thetayloredman** [comment]
@grazzolini Stupid Manjaro installer teaching me I needed 8MB for it... (I originally would make it some 512MB like my EFI-SPs)

**Comment by kevr** [comment]
Just want to leave this note here... it is understandable if folks do not want to implement support for this, however, there is a bit of a misleading issue in the script currently. If you run the guided installation on a system that does not have EFI, the process ends up silently failing and printing out success messages to the user.

Would probably be relatively easy to drop a warning or error in the case where they continue without an EFI-capable system. I'm down to contribute a patch. Thoughts?

**Comment by Torxed** [comment]
> Just want to leave this note here... it is understandable if folks do not want to implement support for this, however, there is a bit of a misleading issue in the script currently. If you run the guided installation on a system that does not have EFI, the process ends up silently failing and printing out success messages to the user.
> 
> Would probably be relatively easy to drop a warning or error in the case where they continue without an EFI-capable system. I'm down to contribute a patch. Thoughts?
> 
> EDIT: Scratch that -- [8355bbe](https://github.com/archlinux/archinstall/commit/8355bbe476a8e0687101222ca3fe592fc91272cc)

Yea it's an err on my side, I assumed people read the Wiki about archinstall before blindly executing a binary/script that some random person wrote ^^ I appreciate people blindly trusting my code, which is why I decided to add that warning at the start of the guided installer. Hopefully we can remove that in `v2.2.0` once MBR/Bios/GRUB support comes out :)

**Comment by ameeno** [comment]
the other issue here is I use an Arch Linux Net install via ipxe. (For virtual machines) most VM's either MS Azure or KVM tend to prefer legacy boot.
IPXE doesn't work too well via uefi.
So I need to start a VM in legacy mode, to boot ipxe arch, to run arch install script to switch to UEFI to boot arch.
pretty cumbersome I think?

In the world of VM's I think legacy boot install is a must.

**Comment by Torxed** [comment]
> the other issue here is I use an Arch Linux Net install via ipxe. (For virtual machines) most VM's either MS Azure or KVM tend to prefer legacy boot.
> IPXE doesn't work too well via uefi.
> So I need to start a VM in legacy mode, to boot ipxe arch, to run arch install script to switch to UEFI to boot arch.
> pretty cumbersome I think?
> 
> In the world of VM's I think legacy boot install is a must.

I PXE with UEFI without any issues, but this isn't always an option in most cloud providers - that's true.
I intend to do more PXE tests in the future, I've also written a few tests cases that I haven't pushed yet for it already with [a side project](https://github.com/Torxed/ArchPXE) for this very use case.. But legacy support will come in one shape or another and use cases like this is valuable feedback! :)

**Comment by ameeno** [comment]
Thanks for your reply.
I have had so much Trouble with UEFI+PXE i just gave up with it.

I know in THEORY Uefi + PXE should work. but in my home setup with a Mikrotik router tftp server with netboot.xyz payload, I have never actually been able to gett uefi + pxe to work ever.
I have heard rumours that in a magical land somewhere outside of my realm of knowledge it *should* be possible to pxe UEFI. But either my desktop Mobo, My laptops, My proxmox VM's my KVM VM's (using bridge networking) or hyper-v VM's have ever liked to PXE over EFI... (It could be my router, as I do not have a dedicated TFTP server/pxe boot server, just using TFTP and dhcp options feature of my router)

As a result, I pretty much always use Legacy boot for VM's. unless I have a very specific usecase like VFIO or something specific.

So yea, In my use case, while the archinstall script is super awesome, it really cannot be used without a legacy boot option :)

Thanks for the valuable script it is awesome whenever I have tested it.

**Comment by Torxed** [comment]
> Thanks for your reply.
> I have had so much trouble with UEFI+PXE I just gave up with it.
> 
> I know in THEORY Uefi + PXE should work. but in my home setup with a Mikrotik router TFTP server with netboot.xyz payload, I have never actually been able to get UEFI + pxe to work ever.
> I have heard rumours that in a magical land somewhere outside of my realm of knowledge it _should_ be possible to pxe UEFI. But either my desktop Mobo, My laptops, My proxmox VM's my KVM VM's (using bridge networking) or hyper-v VM's have ever liked to PXE over EFI... (It could be my router, as I do not have a dedicated TFTP server/pxe boot server, just using TFTP and DHCP options feature of my router)
> 
> As a result, I pretty much always use the Legacy boot for VM's. unless I have a very specific use-case like VFIO or something specific.
> 
> So yea, In my use case, while the archinstall script is super awesome, it really cannot be used without a legacy boot option :)
> 
> Thanks for the valuable script it is awesome whenever I have tested it.

For what it's worth, I'll leave [this script here](https://github.com/Torxed/Scripts/blob/master/bash/qemu.sh#L58) which enables PXE over UEFI for Qemu (kvm):
```bash
qemu-system-x86_64 ... \
    -global i82557b.romfile="efi-eepro100.rompxe"
```
I appreciate the kind words and I'll probably post a comment here even after I've closed this ticket when I get some free time to try out the PXE features :)

**Comment by Torxed** [comment]
I'm considering this issue closed since we do have GRUB and MBR support in master, which will be the next release.
